### ISSUE TITLE
Add environment variables for functions to Lambda configuration

### ISSUE THREAD

**Issue by timmattison** [awslabs/aws-greengrass-provisioner#310]
When a function is created with GGP several environment variables are populated that provide the core's thing name and other important variables. These variables are put in the Greengrass function definition.

These variables should be added to the Lambda configuration for two reasons:

~~1) To make cloud debugging possible with functions that lend themselves to running in the cloud and Greengrass~~
2) To allow functions to be reused in a future release by simply specifying the function ARN. Currently if that feature was added the environment variables would be missing which would cause many functions to break.

**Comment by timmattison** [comment]
Reason 1 may need to be deprioritized. It seems like the best way to support 2 is to create the environment variables in Lambda's environment with a "GGP_" prefix and leave in the variables that need to be replaced. For example, ${THING_NAME}. This way when another group uses the function it can correct replace the variables with its own values.

This means we will:
- Clear out all "GGP_" prefixed variables each time GGP builds a new function
- Put the environment variables in the Lambda function's environment with the "GGP_" prefix without replacing the variables
- Pull the variables out of the Lambda function's environment, replace the variables in the "GGP_" prefixed ones, and then strip the "GGP_" prefix before they're added to the function definition.
- Include any existing environment variables from Lambda in the function definition.

**Comment by timmattison** [comment]
Reason 1 deprioritized

**Comment by timmattison** [comment]
Will start on this today after #263 is done

**Comment by timmattison** [comment]
This requires some moving around of functionality related to publishing Lambda functions and creating versions. BasicFunctionHelper's createFunction method would need to return just the FunctionConf and the LambdaFunctionArnInfoAndFunctionConf object would need to be created elsewhere.

It looks like the call to publishLambdaFunctionVersion (in the working branch) is what needs to be done elsewhere. This shouldn't be too hard since it is reused for all runtimes.

Step 1 is to make createFunctionIfNecessary return void and trace up where that changes the flow. The publish part needs to be placed somewhere where we can add the environment variables and then publish the version for all runtimes (no per language specific code).

**Comment by timmattison** [comment]
Pulling out some of this work into another issue

**Comment by timmattison** [comment]
Prototype is working on a private branch. Needs some cleanup but it is working in initial tests. More integration tests will need to be built to test this.

A different approach was taken than the one above. The current approach is to take the function.conf and store it in AWS Lambda's environment variables. When a separate deployment references an existing function it looks for this variable, pulls it back to GGP, and then rehydrates it as if the function.conf was local.

Then some work is done to make sure it avoids attempting to build or update the function and create a new alias. And finally it is merged together with the other functions, the function definitions are built, and the deployment goes on as it normally would.

**Comment by timmattison** [comment]
Complete.
